kobotoolbox/kpi
# feat(backend): generate OpenAPI v3 swagger docs

### 🗒️ Checklist

1. [x] run linter locally
2. [x] update all related docs (API, README, inline, etc.), if any
3. [x] draft PR with a title `<type>(<scope>)<!>: <title> TASK-1234`
4. [x] tag PR: at least `frontend` or `backend` unless it's global
5. [x] fill in the template below and delete template comments
6. [x] review thyself: read the diff and repro the preview as written
7. [x] open PR & confirm that CI passes
8. [x] request reviewers, if needed
9. [ ] delete this section before merging

### 📣 Summary

Automatically generate and display API schema.

### 📖 Description

Automatically generate OpenAPI v3 schema and display it in both swagger and redoc. See urls:
- /api/schema/swagger
- /api/schema/redoc

### 👀 Preview steps

1. ℹ️ note that python dependencies has changed
2. open /api/schema
3. 🟢 notice that a schema downloads
4. open /api/schema/swagger
5. 🟢 notice that it's a swagger docs of the API schema
6. open /api/schema/redoc
7. 🟢 notice that it's a redoc docs of the API schema

### 💭 Notes

The schema most likely is not complete nor accurate. The scope of this PR is auto-generate schema as-is as a starting point, to enable iteration and improvement of the schema over time.

The purpose of auto-generated schema is twofold:
- use it as basis for auto-generating API types in typescript for the frontend. That's out of scope for another PR.
- a critical part of the automated testing architecture, see [internal proposal here](https://www.notion.so/kobotoolbox/2024-11-05-Automated-Test-Architecture-Options-1357e515f654804cbe98e70c5466ab2c).

I considered three options for schema generation:
- :warning: [django-rest-swagger](https://github.com/marcgibbons/django-rest-swagger) - deprecated since 2019.
- :warning: [drf-yasg](https://github.com/axnsan12/drf-yasg) - maintained, but generates only OpenAPI v2 without intentions to support v3.
- :heavy_check_mark: [drf-spectacular](https://github.com/tfranzel/drf-spectacular) - maintained & support OpenAPIv3

Reviewers, please help me to pack this properly for merging, this PR is rough like a quick PoC.

Screenshots of swagger and redoc:

![image](https://github.com/user-attachments/assets/22699437-da49-4c8d-bb51-91d0831844a5)
![image](https://github.com/user-attachments/assets/c935735f-2b89-4607-8a2f-53722116eec6)
